### PR TITLE
KAFKA-14936: Add Grace period logic to Stream Table Join (2/N)

### PR DESCRIPTION
**Pull request by wcarlson5**
Working PR for KIP-923. Will retarget once 1/N is merged.

This Pr adds the interface for grace period to the Joined object as well as uses the buffer. The majority of it is tests and moving some of the existing join logic.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java [hunk 1]
```diff
@@ -39,15 +47,26 @@
...
+    private final Optional<TimeOrderedKeyValueBuffer<K1, V1>> buffer;
+    protected long observedStreamTime = ConsumerRecord.NO_TIMESTAMP;
+    private InternalProcessorContext internalProcessorContext;
 
     KStreamKTableJoinProcessor(final KTableValueGetter<K2, V2> valueGetter,
                                final KeyValueMapper<? super K1, ? super V1, ? extends K2> keyMapper,
                                final ValueJoinerWithKey<? super K1, ? super V1, ? super V2, ? extends VOut> joiner,
-                               final boolean leftJoin) {
+                               final boolean leftJoin,
+                               final Optional<Duration> gracePeriod,
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:59]: Is there a difference between grace period being empty, and grace period being zero? If not, can we simply this by removing the optional and defaulting grace period to zero instead?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 2]
```diff
@@ -1256,10 +1261,21 @@ private <VO, VR> KStream<K, VR> doStreamTableJoin(final KTable<K, VO> table,
         final NamedInternal renamed = new NamedInternal(joinedInternal.name());
 
         final String name = renamed.orElseGenerateWithPrefix(builder, leftJoin ? LEFTJOIN_NAME : JOIN_NAME);
+
+        Optional<TimeOrderedKeyValueBuffer<K, V>> buffer = Optional.empty();
+
+        if (joined.gracePeriod() != null) {
+            final RocksDBTimeOrderedKeyValueSegmentedBytesStore store = new RocksDbTimeOrderedKeyValueBytesStoreSupplier(name,  1).get();
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:1268]: Why is this retention period always set to 1?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java [hunk 3]
```diff
@@ -39,15 +47,26 @@
...
+                               final Optional<TimeOrderedKeyValueBuffer<K1, V1>> buffer) {
         this.valueGetter = valueGetter;
         this.keyMapper = keyMapper;
         this.joiner = joiner;
         this.leftJoin = leftJoin;
+        if (gracePeriod.isPresent() ^ buffer.isPresent()) {
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:65]: Related to the above, should we simplify the case where grace period is zero to not create/pass a buffer, since it's unnecessary?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java [hunk 4]
```diff
@@ -56,10 +75,58 @@ public void init(final ProcessorContext<K1, VOut> context) {
         final StreamsMetricsImpl metrics = (StreamsMetricsImpl) context.metrics();
         droppedRecordsSensor = droppedRecordsSensor(Thread.currentThread().getName(), context.taskId().toString(), metrics);
         valueGetter.init(context);
+        if (valueGetter.isVersioned() && !gracePeriod.isPresent()) {
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:78]: This logic looks backwards, which makes me think we might be missing test coverage -- it looks like it's supposed to be checking what happens if grace period is present but the value getter is not versioned, instead of checking if the value getter is versioned but grace period is not present.

Also, I think it'd be more user friendly to check for this error when building the topology, rather than when starting the app. We should have enough information to perform the check at the time of building the topology, but if you prefer to do it here / improve this behavior in a follow-up PR, that's fine too.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java [hunk 5]
```diff
@@ -56,10 +75,58 @@ public void init(final ProcessorContext<K1, VOut> context) {
...
+            throw new IllegalArgumentException("KTable must be versioned to use a grace period in a stream table join.");
+        }
+        if (buffer.isPresent()) {
+            internalProcessorContext = asInternalProcessorContext((org.apache.kafka.streams.processor.ProcessorContext) context);
+            buffer.get().setSerdesIfNull(new SerdeGetter(context));
+            //cast doesn't matter, it is just because the processor is deprecated. The context gets converted back with StoreToProcessorContextAdapter.adapt(context)
+            buffer.get().init((org.apache.kafka.streams.processor.StateStoreContext) context(), null);
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:85]: Is passing null here valid? I'm surprised the context is able to register with a null root.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java [hunk 6]
```diff
@@ -56,10 +75,58 @@ public void init(final ProcessorContext<K1, VOut> context) {
...
+            if( record.key() == null) {
+                LOG.warn(
+                    "Skipping record due to null join key or value. "
+                        + "topic=[{}] partition=[{}] offset=[{}]",
+                    internalProcessorContext.recordContext().topic(), internalProcessorContext.recordContext().partition(), internalProcessorContext.recordContext().offset());
+                return;
+            }
+            updateObservedStreamTime(record.timestamp());
+            final long deadline = observedStreamTime - gracePeriod.get().toMillis();
+            if( record.timestamp() < deadline) {
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:103]: Looks like this can be `<=` instead?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java [hunk 7]
```diff
@@ -56,10 +75,58 @@ public void init(final ProcessorContext<K1, VOut> context) {
...
+                doJoin(record);
+            } else {
+                final Change<V1> change = new Change<>(record.value(), record.value());
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:106]: This comment can be ignored if we plan to move away from `Change` here entirely (per discussion on earlier PR), but I think it'd be better to set `oldValue = null` if we're not using `oldValue`, in order to keep the serialization smaller.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java [hunk 8]
```diff
@@ -56,10 +75,58 @@ public void init(final ProcessorContext<K1, VOut> context) {
...
+                final Record<K1, Change<V1>> r = new Record<>(record.key(), change, record.timestamp());
+                buffer.get().put(observedStreamTime, r, internalProcessorContext.recordContext());
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:108]: What's the purpose in storing `internalProcessorContext` separately, instead of calling `context()` as the `doJoin()` method does? 

Also, it looks like the `put(time, record, context)` interface is maybe meant to treat the passed `time` as the timestamp associated with the record, rather than fetching the timestamp from the record itself -- at least that's what the existing in-memory implementation of this interface appears to do. Is that what you're seeing too? The fact that the new RocksDB implementation of this interface ignores the `time` parameter is very confusing/inconsistent.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java [hunk 9]
```diff
@@ -56,10 +75,58 @@ public void init(final ProcessorContext<K1, VOut> context) {
...
+
+    @Override
+    public void close() {
+        valueGetter.close();
+    }
+
+    private void emit(final TimeOrderedKeyValueBuffer.Eviction<K1, V1> toEmit) {
+        final Record<K1, V1> record = new Record<>(toEmit.key(), toEmit.value().newValue, toEmit.recordContext().timestamp())
+            .withHeaders(toEmit.recordContext().headers());
+        doJoin(record);
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:122]: I think we should update the context to be the emitted record's context before performing the join, else downstream calls might use an incorrect context?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java [hunk 10]
```diff
@@ -56,10 +75,58 @@ public void init(final ProcessorContext<K1, VOut> context) {
...
+            buffer.get().init((org.apache.kafka.streams.processor.StateStoreContext) context(), null);
+        }
     }
 
     @Override
     public void process(final Record<K1, V1> record) {
+        if (!gracePeriod.isPresent() || !buffer.isPresent()) {
+            doJoin(record);
+        } else {
+            if( record.key() == null) {
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:94]: This check for null key is checking the original record key, and not the mapped record key. I think we should be checking the mapped record key, and also the record value, like what `doJoin()` does instead?

**Review by vcrfxia — COMMENTED** [review]
Hey @wcarlson5 , left a bunch of questions inline to clarify my understanding of the behavior / with notes on potential bugs. Will have a look at test coverage after we're on the same page about the questions, to save confusion in case my understanding on some counts is wrong.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java (hunk 7 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:106]: Will remove entirely when I rebase on to https://github.com/apache/kafka/pull/13830/files once merged

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java (hunk 5 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:85]: null is ignored and the wrapped store in the buffer is passed in instead. https://github.com/apache/kafka/blob/755f650b5d177f62c366210ee1db5ac71a05aea7/streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedKeyValueBuffer.java#L73

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java (hunk 6 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:103]: sure, it can

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java (hunk 8 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:108]: It looks like to preserve the record's headers. 

I agree the time parameter is a bit confusing. I'm not sure why the in-memory used it as the context as the time as well. In fact it stores the time as part of the key, duplicating the storage costs. I don't think we need it at all, but if we have it pass in observed time we can have it advance the stores time beyond records in the store if we don't want to add those records to the store.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java (hunk 9 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:122]: Yeah, we should set the record context

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java (hunk 10 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:94]: good call

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java (hunk 3 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:65]: That shouldn't ever happen

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java [hunk 11]
```diff
@@ -56,10 +76,74 @@ public void init(final ProcessorContext<K1, VOut> context) {
         final StreamsMetricsImpl metrics = (StreamsMetricsImpl) context.metrics();
         droppedRecordsSensor = droppedRecordsSensor(Thread.currentThread().getName(), context.taskId().toString(), metrics);
         valueGetter.init(context);
+        internalProcessorContext = asInternalProcessorContext((org.apache.kafka.streams.processor.ProcessorContext) context);
+        if (buffer.isPresent()) {
+            if (!valueGetter.isVersioned() && gracePeriod.isPresent()) {
+                throw new IllegalArgumentException("KTable must be versioned to use a grace period in a stream table join.");
+            }
+            if (valueGetter instanceof VersionedBytesStoreSupplier
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:84]: Is this if-condition ever true? I'm not seeing KTableValueGetter implementations which implement VersionedBytesStoreSupplier.

Would be good to get test coverage for this exception message, could be in a follow-up PR.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java [hunk 12]
```diff
@@ -56,10 +76,74 @@ public void init(final ProcessorContext<K1, VOut> context) {
...
+
+    @Override
+    public void close() {
+        valueGetter.close();
+    }
+
+    private void emit(final TimeOrderedKeyValueBuffer.Eviction<K1, V1> toEmit) {
+        final Record<K1, V1> record = new Record<>(toEmit.key(), toEmit.value(), toEmit.recordContext().timestamp())
+            .withHeaders(toEmit.recordContext().headers());
+        internalProcessorContext.setRecordContext(toEmit.recordContext());
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:126]: After setting the record context here, we have to unset it afterwards, right? Would be good to get unit test coverage here since it's a tricky piece of logic.

File: streams/src/test/java/org/apache/kafka/streams/integration/StreamTableJoinWithGraceIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+    @Before
+    public void prepareTopology() throws InterruptedException {
+        super.prepareEnvironment();
+        appID = "stream-table-join-integration-test";
+        builder = new StreamsBuilder();
+        joined = Joined.with(Serdes.Long(), Serdes.String(), Serdes.String(), "Grace", Duration.ZERO);
+    }
+
+    @Test
+    public void testInner() {
```
Comment (vcrfxia) [inline streams/src/test/java/org/apache/kafka/streams/integration/StreamTableJoinWithGraceIntegrationTest.java:70]: This test doesn't look like it tests anything related to your new feature? Seems to be a dupe of the existing test from `StreamTableJoinIntegrationTest`. Same for the one below as well.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinWithGraceTest.java [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+        final KStream<Integer, String> stream;
+        final KTable<Integer, String> table;
+
+        final Consumed<Integer, String> consumed = Consumed.with(Serdes.Integer(), Serdes.String());
+        stream = builder.stream(streamTopic, consumed);
+        table = builder.table(tableTopic, consumed, Materialized.as(
+            Stores.persistentVersionedKeyValueStore("V-grace", Duration.ofMinutes(5))));
+        stream.join(table,
+            MockValueJoiner.TOSTRING_JOINER,
+            Joined.with(Serdes.Integer(), Serdes.String(), Serdes.String(), "Grace", Duration.ZERO)
```
Comment (vcrfxia) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinWithGraceTest.java:83]: Can we add tests which use nonzero grace period as well? All the new tests in this PR look to be testing the zero grace period case. 

Also, I don't see tests in this PR which validate the new functionality provided by the grace period feature. We seem to only be testing that adding grace period as a join argument doesn't break existing functionality of the join?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 15]
```diff
@@ -1256,10 +1261,26 @@ private <VO, VR> KStream<K, VR> doStreamTableJoin(final KTable<K, VO> table,
...
         final String name = renamed.orElseGenerateWithPrefix(builder, leftJoin ? LEFTJOIN_NAME : JOIN_NAME);
+
+        Optional<TimeOrderedKeyValueBuffer<K, V, V>> buffer = Optional.empty();
+
+        if (joined.gracePeriod() != null) {
+            if (!((KTableImpl<K, ?, VO>) table).graphNode.isOutputVersioned().orElse(true)) {
+                throw new RuntimeException("KTable must be versioned to use a grace period in a stream table join.");
+            }
+            final String bufferStoreName = name + "-Buffer";
+            final RocksDBTimeOrderedKeyValueBytesStore store = new RocksDbTimeOrderedKeyValueBytesStoreSupplier(bufferStoreName,
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:1272]: nit: the second argument here looks unused, and can be removed.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java (hunk 5 above)
Comment (vcrfxia) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:90]: That's from your first PR, right? Relates to my other comment about where the responsibility for setting `root` should be: https://github.com/apache/kafka/pull/13756#discussion_r1228495958

**Review by vcrfxia — COMMENTED** [review]
Hey @wcarlson5 , code mostly looks good but I think test coverage for the new feature is lacking. Comments inline -- happy to provide more concrete suggestions for test coverage if that'd be helpful.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java (hunk 12 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinProcessor.java:126]: I dint't think so. The next record will override it and it should get reset if process is called again. But Ill see if there is a good way to unset it

File: streams/src/test/java/org/apache/kafka/streams/integration/StreamTableJoinWithGraceIntegrationTest.java (hunk 13 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/integration/StreamTableJoinWithGraceIntegrationTest.java:70]: oh yeah, this was for my sanity to make sure I didn't mess up the normal join. I can remove that now
